proof/invariant-abstract/DetSchedSchedule_AI.thy

```diff
@@ -1460,11 +1460,20 @@ lemma set_tcb_sched_context_valid_sched_pred':
 lemmas set_tcb_sched_context_valid_sched_pred[valid_sched_wp] =
   set_tcb_sched_context_valid_sched_pred'[THEN hoare_drop_assertion]
 
-lemma set_tcb_sched_context_valid_ready_qs_not_queued:
-  "\<lbrace>valid_ready_qs and not_queued ref\<rbrace>
+lemma set_tcb_sched_context_valid_ready_qs:
+  "\<lbrace>valid_ready_qs and (\<lambda>s. in_ready_q ref s \<longrightarrow> (\<exists>p. sc = Some p \<and> released_sc_at p s))\<rbrace>
    set_tcb_obj_ref tcb_sched_context_update ref sc
    \<lbrace>\<lambda>_. valid_ready_qs\<rbrace>"
-  by (wpsimp wp: valid_sched_wp simp: valid_ready_qs_def vs_all_heap_simps not_queued_def)
+  unfolding set_tcb_obj_ref_def
+  apply simp
+  apply (rule hoare_seq_ext[OF _ gets_the_sp])
+  apply (wpsimp simp: vs_all_heap_simps obj_at_def set_object_def
+                  wp: get_object_wp split: option.splits)
+  apply (cases sc; simp)
+   apply (clarsimp simp: valid_ready_qs_def vs_all_heap_simps not_queued_def dest!: get_tcb_SomeD)+
```

Comment:
nitpick: is the `+` necessary here?

Reply:
I believe so. It discharges the first subgoal and then applies to the second. Can repeat instead.